src/blenderbim/blenderbim/tool/georeference.py

```diff
@@ -0,0 +1,231 @@
+    
+    @classmethod
+    def set_true_north_prop(cls, data):
+        if data:
+            bpy.context.scene.BIMGeoreferenceProperties.true_north_abscissa = str(data[0])
+            bpy.context.scene.BIMGeoreferenceProperties.true_north_ordinate = str(data[1])
+    
+    @classmethod
+    def get_true_north_props(cls, props):
+        true_north = None
+        if props.has_true_north:
+            try:
+                true_north = [float(props.true_north_abscissa), float(props.true_north_ordinate)]
+            except ValueError:
+                self.report({"ERROR"}, "True North Abscissa and Ordinate expect a number") #TODO use the correct report method
+        return true_north
+    
+    @classmethod
+    def enable_editing(cls):
+        bpy.context.scene.BIMGeoreferenceProperties.is_editing = True
+    
+    @classmethod
+    def disable_editing(cls):
+        bpy.context.scene.BIMGeoreferenceProperties.is_editing = False
+    
+    @classmethod
+    def get_file(cls):
+        return tool.Ifc.get()
+    
+    @classmethod
+    def import_projected_crs_attributes(cls, name, prop, data):
+        if name == "MapUnit":
+            new = bpy.context.scene.BIMGeoreferenceProperties.projected_crs.add()
+            new.name = name
+            new.data_type = "enum"
+            new.is_null = data[name] is None
+            new.is_optional = True
+            new.enum_items = json.dumps(
+                {u["id"]: u["Name"] for u in UnitData.units.values() if u["UnitType"] == "LENGTHUNIT"}
+            )
+            if data["MapUnit"]:
+                new.enum_value = str(data["MapUnit"]["id"])
+            return True
+    
+    @classmethod
+    def import_map_conversion_attributes(cls, name, prop, data):
+        if name not in ["SourceCRS", "TargetCRS"]:
+            # Enforce a string data type to prevent data loss in single-precision Blender props
+            prop.data_type = "string"
+            prop.string_value = "" if prop.is_null else str(data[name])
+            return True
+    
+    @classmethod
+    def export_crs_attributes(cls, attributes, prop):
+        ifc_file = tool.Ifc.get()
+        if not prop.is_null and prop.name == "MapUnit":
+            attributes[prop.name] = ifc_file.by_id(int(prop.enum_value))
+            return True
+    
+    @classmethod        
+    def export_map_attributes(cls, attributes, prop):
+        if not prop.is_null and prop.data_type == "string":
+            # We store our floats as string to prevent single precision data loss
+            attributes[prop.name] = float(prop.string_value)
+            return True
+    
+    @classmethod
+    def set_ifc_grid_north(cls):
+        x_angle = -bpy.context.scene.sun_pos_properties.north_offset
+        bpy.context.scene.BIMGeoreferenceProperties.map_conversion.get("XAxisAbscissa").string_value = str(cos(x_angle))
+        bpy.context.scene.BIMGeoreferenceProperties.map_conversion.get("XAxisOrdinate").string_value = str(sin(x_angle))
```

Comment:
So it sounds crazy why 3 lines of code should be split into three locations, but trust me on this one :) The core should be abstract, so perhaps something like:

def set_ifc_grid_north(cls, georeference):
    georeference.set_xaxis_vector_from_angle(georeference.get_x_angle_from_sun_position())

So the x_angle in the future could come from other sources like a Cycles sun object, these functions could be reused in another operator (e.g. auto generate shadow diagram operator), and so on. So it sounds overkill but it pays off :)

Reply:
Ok, i follow you. I'll try and i'll tell you if i can't make it.

Reply:
Ok, i'll fix. I try to simplify also other methods in the same way (for example def set_blender_grid_north(georeference) in core)